ztools/ztls/handshake_client.go

```diff
@@ -32,19 +34,255 @@ type clientHandshakeState struct {
+		return nil, errors.New("tls: Too many compression methods")
+	}
+	compressions := make([]byte, len(c.CompressionMethods)+1)
+	compressions[0] = uint8(len(c.CompressionMethods))
+	if len(c.CompressionMethods) > 0 {
+		copy(compressions[1:], c.CompressionMethods)
+		if c.CompressionMethods[0] != 0 {
+			return nil, errors.New(fmt.Sprintf("tls: unimplemented compression method %d", c.CompressionMethods[0]))
+		}
+		if len(c.CompressionMethods) > 1 {
+			return nil, errors.New(fmt.Sprintf("tls: unimplemented compression method %d", c.CompressionMethods[1]))
+		}
+	} else {
+		return nil, errors.New("tls: no compression method")
+	}
+
+	var extensions []byte
+	for _, ext := range c.Extensions {
+		extensions = append(extensions, ext.Marshal()...)
+	}
+	if len(extensions) > 0 {
+		length := make([]byte, 2)
+		length[0] = uint8(len(extensions) >> 8)
+		length[1] = uint8(len(extensions))
+		extensions = append(length, extensions...)
+	}
+	helloArray := [][]byte{head, sessionID, ciphers, compressions, extensions}
+	hello := []byte{}
+	for _, b := range helloArray {
+		hello = append(hello, b...)
+	}
+	lengthOnTheWire := len(hello) - 4
+	if lengthOnTheWire >= 1<<24 {
+		return nil, errors.New("ClientHello message too long")
+	}
+	hello[1] = uint8(lengthOnTheWire >> 16)
+	hello[2] = uint8(lengthOnTheWire >> 8)
+	hello[3] = uint8(lengthOnTheWire)
+
+	return hello, nil
+}
+
 func (c *Conn) clientHandshake() error {
 	if c.config == nil {
 		c.config = defaultConfig()
 	}
-
 	var hello *clientHelloMsg
-
+	var helloBytes []byte
 	var session *ClientSessionState
 	var sessionCache ClientSessionCache
 	var cacheKey string
 
-	// first, let's check if a ClientHello template was provided by the user
-	if c.config.ExternalClientHello != nil {
+	if c.config.ClientFingerprintConfiguration != nil {
+		if err := c.config.ClientFingerprintConfiguration.WriteToConfig(c.config); err != nil {
+			return err
+		}
+	}
+
+	// first, let's check if a ClientFingerprintConfiguration template was provided by the config
+	if c.config.ClientFingerprintConfiguration != nil {
+		session = nil
+		sessionCache = c.config.ClientFingerprintConfiguration.SessionCache
+		if sessionCache != nil {
+			if c.config.ClientFingerprintConfiguration.CacheKey == nil {
+				return errors.New("Must specify CacheKey if SessionCache is defined in Config.ClientFingerprintConfiguration")
+			}
+			cacheKey = c.config.ClientFingerprintConfiguration.CacheKey.Key(c.conn.RemoteAddr())
+			candidateSession, ok := sessionCache.Get(cacheKey)
+			if ok {
+				cipherSuiteOk := false
+				for _, id := range c.config.ClientFingerprintConfiguration.CipherSuites {
+					if id == candidateSession.cipherSuite {
+						cipherSuiteOk = true
+						break
+					}
+				}
+				versOk := candidateSession.vers >= c.config.minVersion() &&
+					candidateSession.vers <= c.config.ClientFingerprintConfiguration.HandshakeVersion
+				if versOk && cipherSuiteOk {
+					session = candidateSession
+				}
```

Comment:
Is this not an error case otherwise? Do you have to match version to the previous session if resuming?